{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2020-02-24 11:03:20,532 lifd:37 INFO: VarCode is used for mutation effect prediction.\n"
     ]
    }
   ],
   "source": [
    "import logging\n",
    "import os\n",
    "import pandas as pd\n",
    "from lifd.lifd import LiFD, LIFD_COL, LIFD_SUP_COL, LIFD1_SUP_COL, LIFD2_SUP_COL\n",
    "from lifd.databases.hotspots_database import HotspotsDB\n",
    "from lifd.databases.cgi_database import CgiDB\n",
    "from lifd.databases.oncokb_database import OncoKBDB\n",
    "from lifd.databases.cosmic_db import CosmicDB\n",
    "from lifd.predictors.vep import Vep\n",
    "from lifd.predictors.candra import Candra\n",
    "from lifd.predictors.cravat import Cravat\n",
    "from lifd.predictors.cgi import Cgi\n",
    "from lifd.predictors.fathmm import FatHMM\n",
    "from lifd.settings import HOTSPOTS_FP, ONCOGENIC_VARS_FP, ONCOKB_ALLVARS_FP, COSMIC_VARS_FP, REF_GENOME_FA_FP\n",
    "from lifd.utils import NT_VAR_COL, PT_VAR_COL, FUNC_COL, MUT_EFFECT_COL, init\n",
    "from lifd.mutation_effects import get_mutation_details, annotate_effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2020-02-24 11:03:20,545 lifd.lifd.predictors.cgi:62 INFO: Successfully set new username reiter.j@gmail.com and token.\n",
      "2020-02-24 11:03:20,546 lifd.lifd.utils:44 INFO: Output directory: /Users/reiter/workspaces/LiFD_dev/lifd_examples/LiFD_TMP\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read file with 7 variants: example_variants.xlsx\n"
     ]
    }
   ],
   "source": [
    "from lifd.cgi_settings import CGI_USER_ID, CGI_TOKEN\n",
    "# from lifd.excluded_module.cgi_settings import CGI_USER_ID, CGI_TOKEN\n",
    "# CGI_USER_ID = '<YOUR_CGI_USERNAME>'\n",
    "# CGI_TOKEN = '<YOUR_CGI_TOKEN>'\n",
    "Cgi.set_login(CGI_USER_ID, CGI_TOKEN)\n",
    "\n",
    "logger = logging.getLogger('lifd.{}'.format(__name__))\n",
    "\n",
    "output_dir = init()\n",
    "\n",
    "# excel file with input variants\n",
    "input_fp = 'example_variants.xlsx'\n",
    "# CSV file with list of putative driver genes (e.g. TCGA consensus list, Bailey et al, Cell 2018)\n",
    "driver_gene_fp = 'BaileyDing2018_driverconsensus.csv'\n",
    "\n",
    "from pathlib import Path\n",
    "input_p = Path(input_fp)\n",
    "output_fn = '{}_LiFDed.xlsx'.format(input_p.stem)\n",
    "output_fp = os.path.join(input_p.parent, output_fn)\n",
    "\n",
    "if input_p.suffix == '.xlsx' or input_p.suffix == '.xls':\n",
    "    var_df = pd.read_excel(input_fp)\n",
    "    print('Read file with {} variants: {}'.format(len(var_df), input_fp))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2020-02-24 11:03:20,577 lifd.lifd.mutation_effects:60 INFO: Set reference genome to hg19.\n",
      "2020-02-24 11:03:20,578 lifd:84 INFO: Using databases: Hotspots, CGI_Catalog, OncoKB, COSMIC\n",
      "2020-02-24 11:03:20,579 lifd:98 INFO: Using predictors: Vep\n",
      "2020-02-24 11:03:20,585 lifd:120 INFO: Using 299 driver genes for annotation from /Users/reiter/databases/BaileyDing2018_driverconsensus.csv.\n",
      "2020-02-24 11:03:20,586 lifd:123 INFO: Initialized LiFD 0.1.1 with 4 databases and 1 predictors.\n"
     ]
    }
   ],
   "source": [
    "# set up databases for first phase of LiFD\n",
    "hs_db = HotspotsDB(HOTSPOTS_FP)\n",
    "cgi_db = CgiDB(ONCOGENIC_VARS_FP)\n",
    "oncokb_db = OncoKBDB(ONCOKB_ALLVARS_FP)\n",
    "cosmic_db = CosmicDB(COSMIC_VARS_FP)\n",
    "dbs = [hs_db, cgi_db, oncokb_db, cosmic_db\n",
    "      ]\n",
    "\n",
    "# set up predictors for second phase of LiFD\n",
    "prds = [Vep, #Cravat, # Candra, FatHMM, Cgi\n",
    "       ]\n",
    "# prds = [FatHMM]\n",
    "lifd = LiFD(databases=dbs, predictors=prds, \n",
    "            driver_gene_fp=driver_gene_fp, driver_gene_col='TCGADrClf', )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /Users/reiter/Library/Caches/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.cdna.all.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /Users/reiter/Library/Caches/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.ncrna.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /Users/reiter/Library/Caches/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.pep.all.fa.gz.pickle\n",
      "2020-02-24 11:03:21,383 lifd:219 INFO: Added potentially functional driver gene mutation annotation to 5 variants.\n",
      "2020-02-24 11:03:21,383 lifd:225 INFO: LIFD output directory: /Users/reiter/workspaces/LiFD_dev/lifd_examples/LiFD_TMP\n",
      "2020-02-24 11:03:21,385 lifd.lifd.databases.hotspots_database:49 INFO: Reading database Hotspots from file /Users/reiter/databases/ChangTaylor_hotspots_cd2018_v2.xls.\n",
      "2020-02-24 11:03:21,984 lifd.lifd.databases.hotspots_database:77 INFO: Loaded 2547 cancer hotspot variants from file /Users/reiter/databases/ChangTaylor_hotspots_cd2018_v2.xls.\n",
      "2020-02-24 11:03:21,986 lifd.lifd.databases.cgi_database:54 INFO: Reading database CGI_Catalog from file /Users/reiter/databases/TamboreroLopezBigas_cgi_oncogenic_gm2018.tsv.\n",
      "2020-02-24 11:03:21,990 lifd.lifd.databases.cgi_database:72 INFO: Read previously processed CGI database with 5722 distinct mutations.\n",
      "2020-02-24 11:03:21,991 lifd.lifd.databases.cgi_database:128 INFO: Loaded 5722 CGI oncogenic variants from file /Users/reiter/databases/TamboreroLopezBigas_cgi_oncogenic_gm2018.tsv.\n",
      "2020-02-24 11:03:21,996 lifd.lifd.databases.oncokb_database:74 INFO: Reading database OncoKB from file /Users/reiter/databases/oncoKB_allAnnotatedVariants_v2.1.tsv.\n",
      "2020-02-24 11:03:22,121 lifd.lifd.databases.oncokb_database:90 INFO: Loaded 3920 annotated oncogenic variants in OncoKB from file /Users/reiter/databases/oncoKB_allAnnotatedVariants_v2.1.tsv.\n",
      "2020-02-24 11:03:22,123 lifd.lifd.databases.cosmic_db:58 INFO: Reading database COSMIC from file /Users/reiter/databases/CosmicGenomeScreensMutantExport_grch37_v89.tsv.\n",
      "2020-02-24 11:03:25,272 lifd.lifd.databases.cosmic_db:72 INFO: Read previously processed genome/exome-wide sequencing data of COSMIC samples with in total 6210551 distinct mutations.\n",
      "2020-02-24 11:03:27,034 lifd.lifd.predictors.vep:191 INFO: Generated VEP input file from dataframe with 7 entries: LiFD_TMP/PAAD_vep_input.tsv\n",
      "2020-02-24 11:03:27,035 lifd.lifd.predictors.vep:210 INFO: Call VEP with command ./vep --species homo_sapiens --dir /Users/reiter/.vep --assembly GRCh37 --cache_version 99 --no_progress --offline --no_stats --sift b --symbol --numbers --domains --gene_phenotype --pubmed --protein --variant_class --shift_hgvs 1 --check_existing --no_escape --failed 1 --minimal --allele_number --pick_order canonical,tsl,biotype,rank,ccds,length --input_file /Users/reiter/workspaces/LiFD_dev/lifd_examples/LiFD_TMP/PAAD_vep_input.tsv --output_file /Users/reiter/workspaces/LiFD_dev/lifd_examples/LiFD_TMP/PAAD_vep_output.tsv --polyphen b --buffer_size 50 --force_overwrite --pick\n",
      "2020-02-24 11:03:27,926 lifd.lifd.predictors.vep:218 INFO: Successfully called VEP: /Users/reiter/workspaces/LiFD_dev/lifd_examples/LiFD_TMP/PAAD_vep_output.tsv\n",
      "2020-02-24 11:03:27,946 lifd.lifd.predictors.vep:340 WARNING: Found 2 duplicates in VEP output according to the NtVarKey!\n",
      "2020-02-24 11:03:27,952 lifd.lifd.predictors.vep:344 INFO: Loaded 6 VEP annotated variants from file LiFD_TMP/PAAD_vep_output.tsv.\n",
      "2020-02-24 11:03:27,964 lifd.lifd.utils:125 INFO: Added column PolyPhen with 5 values for 7 entries.\n",
      "2020-02-24 11:03:27,975 lifd.lifd.utils:125 INFO: Added column Sift with 5 values for 7 entries.\n",
      "2020-02-24 11:03:27,984 lifd.lifd.utils:125 INFO: Added column VEP_impact with 7 values for 7 entries.\n",
      "2020-02-24 11:03:27,998 lifd:258 INFO: LiFD predicts that 5 of 5 (100.0%) nonsynonymous or splice-site variants in driver genes are functional.\n",
      "2020-02-24 11:03:28,002 lifd:264 INFO: Positive predictions of LiFD1 4/5 (80.0%)\n",
      "2020-02-24 11:03:28,006 lifd:270 INFO: Positive predictions of LiFD2 1/5 (20.0%)\n",
      "2020-02-24 11:03:28,053 lifd:326 INFO: Exported 7 annotated variants (5 LiFD) to /Users/reiter/workspaces/LiFD_dev/lifd_examples/LiFD_TMP/example_variants_LiFDed.xlsx.\n"
     ]
    }
   ],
   "source": [
    "var_df = lifd.run_lifd(var_df, output_dir=output_dir, export_fn=output_fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>StartPosition</th>\n",
       "      <th>EndPosition</th>\n",
       "      <th>ReferenceAllele</th>\n",
       "      <th>AlternateAllele</th>\n",
       "      <th>Subject</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>CancerType</th>\n",
       "      <th>NtVarKey</th>\n",
       "      <th>PtVarKey</th>\n",
       "      <th>...</th>\n",
       "      <th>In_OncoKB</th>\n",
       "      <th>In_COSMIC</th>\n",
       "      <th>LiFD1_support</th>\n",
       "      <th>PolyPhen</th>\n",
       "      <th>Sift</th>\n",
       "      <th>VEP_impact</th>\n",
       "      <th>LiFD2_results</th>\n",
       "      <th>LiFD2_support</th>\n",
       "      <th>LiFD_support</th>\n",
       "      <th>LiFD</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>P1</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>PAAD</td>\n",
       "      <td>12__25398284__C__A</td>\n",
       "      <td>KRAS__G12V</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>1149</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.972</td>\n",
       "      <td>0.00</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>[0.0, 1.0]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>48994757</td>\n",
       "      <td>48994756</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>P1</td>\n",
       "      <td>LMTK3</td>\n",
       "      <td>PAAD</td>\n",
       "      <td>19__48994757__-__G</td>\n",
       "      <td>LMTK3__E1407fs</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>20</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>[1.0, 1.0]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>17</td>\n",
       "      <td>7573996</td>\n",
       "      <td>7573996</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>P2</td>\n",
       "      <td>TP53</td>\n",
       "      <td>PAAD</td>\n",
       "      <td>17__7573996__A__G</td>\n",
       "      <td>TP53__L344P</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>7</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.967</td>\n",
       "      <td>0.00</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>[0.0, 1.0]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>27058029</td>\n",
       "      <td>27058029</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>P2</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>PAAD</td>\n",
       "      <td>1__27058029__T__G</td>\n",
       "      <td>ARID1A__Y579*</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>[1.0, 1.0]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>12</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>P2</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>PAAD</td>\n",
       "      <td>12__25398284__C__A</td>\n",
       "      <td>KRAS__G12V</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>1149</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.972</td>\n",
       "      <td>0.00</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>[0.0, 1.0]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>119427951</td>\n",
       "      <td>119427951</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>P2</td>\n",
       "      <td>TBX15</td>\n",
       "      <td>PAAD</td>\n",
       "      <td>1__119427951__T__G</td>\n",
       "      <td>TBX15__S405R</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.719</td>\n",
       "      <td>0.05</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>[0.0, 1.0]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>12</td>\n",
       "      <td>25380275</td>\n",
       "      <td>25380275</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>P3</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>PAAD</td>\n",
       "      <td>12__25380275__T__A</td>\n",
       "      <td>KRAS__Q61H</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>50</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.121</td>\n",
       "      <td>0.00</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>[0.0, 1.0]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  StartPosition  EndPosition ReferenceAllele AlternateAllele  \\\n",
       "0         12       25398284     25398284               C               A   \n",
       "1         19       48994757     48994756               -               G   \n",
       "2         17        7573996      7573996               A               G   \n",
       "3          1       27058029     27058029               T               G   \n",
       "4         12       25398284     25398284               C               A   \n",
       "5          1      119427951    119427951               T               G   \n",
       "6         12       25380275     25380275               T               A   \n",
       "\n",
       "  Subject GeneSymbol CancerType            NtVarKey        PtVarKey  ...  \\\n",
       "0      P1       KRAS       PAAD  12__25398284__C__A      KRAS__G12V  ...   \n",
       "1      P1      LMTK3       PAAD  19__48994757__-__G  LMTK3__E1407fs  ...   \n",
       "2      P2       TP53       PAAD   17__7573996__A__G     TP53__L344P  ...   \n",
       "3      P2     ARID1A       PAAD   1__27058029__T__G   ARID1A__Y579*  ...   \n",
       "4      P2       KRAS       PAAD  12__25398284__C__A      KRAS__G12V  ...   \n",
       "5      P2      TBX15       PAAD  1__119427951__T__G    TBX15__S405R  ...   \n",
       "6      P3       KRAS       PAAD  12__25380275__T__A      KRAS__Q61H  ...   \n",
       "\n",
       "  In_OncoKB In_COSMIC LiFD1_support PolyPhen  Sift VEP_impact  LiFD2_results  \\\n",
       "0      True      1149           4.0    0.972  0.00   MODERATE     [0.0, 1.0]   \n",
       "1     False        20           1.0      NaN   NaN       HIGH     [1.0, 1.0]   \n",
       "2      True         7           3.0    0.967  0.00   MODERATE     [0.0, 1.0]   \n",
       "3     False         0           0.0      NaN   NaN       HIGH     [1.0, 1.0]   \n",
       "4      True      1149           4.0    0.972  0.00   MODERATE     [0.0, 1.0]   \n",
       "5     False         0           0.0    0.719  0.05   MODERATE     [0.0, 1.0]   \n",
       "6      True        50           4.0    0.121  0.00   MODERATE     [0.0, 1.0]   \n",
       "\n",
       "   LiFD2_support  LiFD_support   LiFD  \n",
       "0            0.0           4.0   True  \n",
       "1            1.0           2.0  False  \n",
       "2            0.0           3.0   True  \n",
       "3            1.0           1.0   True  \n",
       "4            0.0           4.0   True  \n",
       "5            0.0           0.0  False  \n",
       "6            0.0           4.0   True  \n",
       "\n",
       "[7 rows x 28 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Chromosome', 'StartPosition', 'EndPosition', 'ReferenceAllele',\n",
       "       'AlternateAllele', 'Subject', 'GeneSymbol', 'CancerType',\n",
       "       'NtVarKey', 'PtVarKey', 'Transcript_ID', 'Protein_ID',\n",
       "       'FatHMM_Key', 'MutationEffect', 'MaybeFunctional', 'TCGADrClf',\n",
       "       'In_Hotspots', 'In_CGI_Catalog', 'In_OncoKB', 'In_COSMIC',\n",
       "       'LiFD1_support', 'PolyPhen', 'Sift', 'VEP_impact', 'LiFD2_results',\n",
       "       'LiFD2_support', 'LiFD_support', 'LiFD'], dtype=object)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var_df.columns.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ########################### debugging ############################\n",
    "# import pysam\n",
    "# from lifd.settings import REF_GENOME_FA_FP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genome = pysam.Fastafile(REF_GENOME_FA_FP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genome.fetch('chr5', 12755, 12858)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# curl -H \"Authorization: Bearer [your token]\" https://www.oncokb.org/api/v1/genes\n",
    "\n",
    "# from lifd.oncokb_settings import ONCOKB_TOKEN\n",
    "# import pycurl\n",
    "# from io import BytesIO\n",
    "\n",
    "# buffer = BytesIO()\n",
    "# crl = pycurl.Curl()\n",
    "# crl.setopt(crl.URL, 'https://www.oncokb.org/api/v1/info')\n",
    "# crl.setopt(crl.WRITEDATA, buffer)\n",
    "# crl.perform()\n",
    "# crl.close()\n",
    "\n",
    "# buffer.getvalue()\n",
    "\n",
    "# buffer = BytesIO()\n",
    "# crl = pycurl.Curl()\n",
    "# crl.setopt(crl.URL, 'https://oncokb.org:443/api/v1/utils/allAnnotatedVariants.txt')\n",
    "# crl.setopt(crl.WRITEDATA, buffer)\n",
    "# crl.perform()\n",
    "# crl.close()\n",
    "\n",
    "# body = buffer.getvalue()\n",
    "# body\n",
    "\n",
    "# with open('test.tsv', 'wb') as f:\n",
    "#     f.write(buffer.getvalue())\n",
    "# buffer.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
